app/celery/broadcast_message_tasks.py

```diff
@@ -1,17 +1,93 @@
 import uuid
+from datetime import datetime
 
 from flask import current_app
 from notifications_utils.statsd_decorators import statsd
 from sqlalchemy.schema import Sequence
+from celery.exceptions import MaxRetriesExceededError
 
 from app import cbc_proxy_client, db, notify_celery
+from app.clients.cbc_proxy import CBCProxyFatalException, CBCProxyRetryableException
 from app.config import QueueNames
-from app.models import BroadcastEventMessageType, BroadcastProvider
-from app.dao.broadcast_message_dao import dao_get_broadcast_event_by_id, create_broadcast_provider_message
+from app.models import BroadcastEventMessageType, BroadcastProvider, BroadcastProviderMessageStatus
+from app.dao.broadcast_message_dao import (
+    dao_get_broadcast_event_by_id,
+    create_broadcast_provider_message,
+    update_broadcast_provider_message_status
+)
 
 from app.utils import format_sequential_number
 
 
+def get_retry_delay(retry_count):
+    """
+    Given a count of retries so far, return a delay for the next one.
+    `retry_count` should be 0 the first time a task fails.
+    """
+    # TODO: replace with celery's built in exponential backoff
+
+    # 2 to the power of x. 1, 2, 4, 8, 16, 32, ...
+    delay = 2**retry_count
+    # never wait longer than 5 minutes
+    return min(delay, 300)
+
+
+def check_provider_message_should_send(broadcast_event, provider):
+    """
+    If any previous event hasn't sent yet for that provider, then we shouldn't send the current event. Instead, fail and
+    raise a P1 - so that a notify team member can assess the state of the previous messages, and if necessary, can
+    replay the `send_broadcast_provider_message` task if the previous message has now been sent.
+
+    Note: This is called before the new broadcast_provider_message is created.
+
+    # Help, I've come across this code following a pagerduty alert, what should I do?
+
+    1. Find the failing broadcast_provider_message associated with the previous event that caused this to trip.
+    2. If that provider message is still failing to send, fix the issue causing that. The task to send that previous
+       message might still be retrying in the background - look for logs related to that task.
+    3. If that provider message has sent succesfully, you might need to send this task off depending on context. This
+       might not always be true though, for example, it may not be necessary to send a cancel if the original alert has
+       already expired.
+    4. If you need to re-send this task off again, you'll need to run the following command on paas:
+       `send_broadcast_provider_message.apply_async(args=(broadcast_event_id, provider), queue=QueueNames.BROADCASTS)`
+    """
+    if broadcast_event.transmitted_finishes_at < datetime.utcnow():
+        # TODO: This should be a different kind of exception to distinguish "We should know something went wrong, but
+        # no immediate action" from "We need to fix this immediately"
```

Comment:
Does this comment still stand? Is the CBCProxyFatalException not a way to show that this is something needs fixing immediately?

Reply:
yeah, this comment still stands, but i haven't perhaps explained myself quite right. This particular case doesn't require immediate action, as the alert has expired naturally so whatever message we were trying to send out no longer needs to be sent out. We should investigate why it didn't send the next day, but we don't necessarily need to get out of bed. 

Compare this case to, eg, an AttributeError - where regardless of what the retry loop does, we need to urgently look into it because a message might need to go out right now.

Perhaps you could make the case that we might still want to get out of bed to understand why a message didn't go out in time, so it doesn't matter what exception type we throw here as long as it's logged. 

But I think it's a bit premature to look into this too much now - I think once some of those code is a bit more stable and we understand a bit more especially around what kind of errors we might get back from proxy I'd like to look into how we log errors and what triggers different pagerduty settings.

Reply:
Gotcha!

That makes sense. Your code comment could be a little clearer to give that context you mentioned but its not the end of the world.

However, I also can't see us hopefully ever having to throw an exception here. On the assumption that at least for a while, alerts won't expire for 24 hours, it would mean that we would get an exception thrown here if we tried to send an alert, it was failing for a while and we couldn't fix it within 24 hours - I hope that never happens.

More likely would be a coding bug that for some reason we were trying to send an old event that had already expired when we shouldn't have.

Reply:
yeah, either that or someone trying to send a cancel at exactly the time it expires anyway cos they expected to have a manual step